File: src/ramatlab/unsupervised/irank.m
%{
    irank - Aggregate rankings using the iRANk algorithm and output results to a CSV file.
    
    This function imports the Python module `irank` and calls its `irank` function to perform
    rank aggregation based on input data provided in a CSV file. The output is also saved in a 
    specified CSV format.

    Reference:
    ----------
    - Wei, F., Li, W., & Liu, S. (2010). iRANK: A rank‐learn‐combine framework for unsupervised ensemble ranking. Journal of the American Society for Information Science and Technology, 61(6), 1232-1243.

    Authors:
    --------
    Qi Deng

    Date:
    -----
    2024-10-11
%}

function irank(input_file_path, output_file_path, input_type)
    % irank - Calls the Python iRANk algorithm for rank aggregation.
    %
    % This function imports the Python module `irank` and calls its `irank` function.
    % It processes the input CSV file containing voting data and outputs the aggregated rankings 
    % to the specified output CSV file.
    %
    % Parameters
    % ----------
    % input_file_path : str
    %     Path to the input CSV file containing voting data. The input should be in CSV format 
    %     with the following columns:
    %     - Query: Does not require consecutive integers starting from 1.
    %     - Voter Name: Allowed to be in string format.
    %     - Item Code: Allowed to be in string format.
    %     - Item Score/Item Rank: Represents the score/rank given by each voter. 
    %
    % output_file_path : str
    %     Path to the output CSV file where the aggregated rankings will be written.
    %
    % input_type : InputType
    %     The type of input data, which determines the interpretation of the fourth column 
    %     as either 'Item Rank' or 'Item Score'.
    
    irankmodule = py.importlib.import_module('irank');
    irankmodule.irank(input_file_path, output_file_path, input_type);
end

File: src/rapython/unsupervised/irank.py
"""
iRANk Algorithm

This implementation is based on the following reference:

Reference:
-----------
- Wei, F., Li, W., & Liu, S. (2010). iRANK: A rank‐learn‐combine framework for unsupervised ensemble ranking. Journal of the American Society for Information Science and Technology, 61(6), 1232-1243.

Authors:
    Qi Deng
Date:
    2024-10-11
"""
import numpy as np

from src.rapython.datatools import *
from src.rapython.common.constant import InputType

__all__ = ['irank']


def irank_agg(sim):
    rankernum = sim.shape[0]
    querynum = sim.shape[1]
    itemnum = sim.shape[2]

    ranklist = np.argsort(-sim, axis=2)
    rank = np.argsort(ranklist, axis=2) + 1

    if itemnum < 10:
        topk = itemnum
    else:
        topk = 10
    superviserank = np.zeros((rankernum - 1, querynum, itemnum))

    for iteration in range(3):
        newsim = sim * 0.9
        for i in range(rankernum):
            if i == 0:
                superviserank = rank[1:rankernum + 1, :, :]
            elif i == rankernum - 1:
                superviserank = rank[0:rankernum, :, :]
            else:
                superviserank[0:i, :, :] = rank[0:i, :, :]
                superviserank[i:rankernum, :, :] = rank[i + 1:rankernum + 1, :, :]

            dscore = 1.0 / superviserank
            # print(dscore.shape)
            dscoretotal = np.sum(dscore, axis=0)
            # print(dscoretotal.shape)
            sresultlist = np.argsort(-dscoretotal, axis=1)

            sresultlist = sresultlist.reshape(querynum, itemnum)

            for k in range(querynum):
                for ll in range(topk):
                    newsim[i, k, sresultlist[k, ll]] += 0.1

        sim = newsim
        ranklist = np.argsort(-sim, axis=2)
        rank = np.argsort(ranklist, axis=2) + 1

    finalsim = np.sum(sim, axis=0)
    # print(finalsim.shape)
    finalsim = finalsim.reshape(querynum, itemnum)
    result = np.argsort(-finalsim, axis=1)
    result = np.argsort(result, axis=1)
    return result


def irank(input_file_path, output_file_path, input_type=InputType.SCORE):
    """
    Process the input CSV file to aggregate rankings and write the results to an output CSV file.
    Parameters
    ----------
    input_file_path : str
        Path to the input CSV file.
        The input to the algorithm should be in CSV file format with the following columns:

        - Query: Does not require consecutive integers starting from 1.
        - Voter Name: Allowed to be in string format.
        - Item Code: Allowed to be in string format.
        - Item Score/Item Rank: Represents the score/rank given by each voter. It is recommended to choose the score format
    output_file_path : str
        Path to the output CSV file.
    input_type : InputType, optional
        The type of input data, defaults to InputType.RANK. It determines
        the naming of the fourth column, which will either be 'Item Rank'
        or 'Item Score' based on this value.
    """
    input_type = InputType.check_input_type(input_type)
    df, unique_queries = csv_load(input_file_path, input_type)
    numpy_data, queries_mapping_dict = df_to_numpy(df, input_type)
    save_as_csv(output_file_path, irank_agg(numpy_data), queries_mapping_dict)


if __name__ == '__main__':
    input_file_loc = r"D:\LocalGit\Agg-Benchmarks\test\full_lists\data\simulation_test.csv"
    output_file_loc = "irank.csv"
    irank(input_file_loc, output_file_loc, InputType.RANK)
